### PR TITLE
[WIP] Add typing for `pytorch_lightning.utilities`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

Part of #7037 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -161,14 +165,14 @@ def batch_to(data):
     return apply_to_collection(batch, dtype=dtype, function=batch_to)
 
 
-def convert_to_tensors(data, device: torch.device = None):
+def convert_to_tensors(data: Any, device: Union[str, torch.device] = None) -> Any:
     if device is None:
         raise MisconfigurationException("device (torch.device) should be provided.")
```
Comment (ananthsub) [inline pytorch_lightning/utilities/apply_func.py:171]: why do we set the device to None as the default and then error if it's not set? 
Should we remove the default value? or type it as Optional?

**Review by ananthsub — APPROVED** [review]
awesome work! i added one comment inline

File: pytorch_lightning/utilities/apply_func.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/apply_func.py:171]: Yes, we could.

File: pytorch_lightning/utilities/cli.py [hunk 2]
```diff
@@ -92,16 +93,16 @@ class LightningCLI:
 
     def __init__(
         self,
-        model_class: Type[LightningModule],
-        datamodule_class: Type[LightningDataModule] = None,
-        save_config_callback: Type[SaveConfigCallback] = SaveConfigCallback,
-        trainer_class: Type[Trainer] = Trainer,
-        trainer_defaults: Dict[str, Any] = None,
-        seed_everything_default: int = None,
+        model_class: 'pl.LightningModule',
```
Comment (tchaton) [inline pytorch_lightning/utilities/cli.py:96]: Type was correct there as we passed a class and not an instance.

File: pytorch_lightning/utilities/apply_func.py [hunk 3]
```diff
@@ -33,13 +33,17 @@
     Batch = type(None)
 
 
-def to_dtype_tensor(value, dtype: torch.dtype = None, device: torch.device = None):
+def to_dtype_tensor(
+    value: Union[int, float, List[Union[int, float]]],
+    dtype: Optional[torch.dtype] = None,
+    device: Union[str, torch.device] = None
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/apply_func.py:39]:
```suggestion
    device: Optional[Union[str, torch.device]] = None,
```

File: pytorch_lightning/utilities/apply_func.py [hunk 4]
```diff
@@ -33,13 +33,17 @@
...
+) -> torch.Tensor:
     if device is None:
         raise MisconfigurationException("device (torch.device) should be provided.")
     return torch.tensor(value, dtype=dtype, device=device)
 
 
-def from_numpy(value, device: torch.device = None):
+def from_numpy(value: np.ndarray, device: Union[str, torch.device] = None) -> torch.Tensor:
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/apply_func.py:46]:
```suggestion
def from_numpy(value: np.ndarray, device: Optional[Union[str, torch.device]] = None) -> torch.Tensor:
```

File: pytorch_lightning/utilities/apply_func.py [hunk 5]
```diff
@@ -56,11 +60,11 @@ def from_numpy(value, device: torch.device = None):
...
     data: Any,
-    dtype: Union[type, tuple],
+    dtype: Union[torch.dtype, Tuple[torch.dtype]],
     function: Callable,
-    *args,
-    wrong_dtype: Optional[Union[type, tuple]] = None,
-    **kwargs
+    *args: Any,
+    wrong_dtype: Optional[Union[torch.dtype, Tuple[torch.dtype]]] = None,
+    **kwargs: Any
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/apply_func.py:67]:
```suggestion
    **kwargs: Any,
```

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -161,14 +165,14 @@ def batch_to(data):
     return apply_to_collection(batch, dtype=dtype, function=batch_to)
 
 
-def convert_to_tensors(data, device: torch.device = None):
+def convert_to_tensors(data: Any, device: Union[str, torch.device] = None) -> Any:
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/apply_func.py:168]:
```suggestion
def convert_to_tensors(data: Any, device: Optional[Union[str, torch.device]] = None) -> Any:
```

File: pytorch_lightning/utilities/cloud_io.py [hunk 7]
```diff
@@ -33,7 +36,7 @@ def load(path_or_url: Union[str, IO, Path], map_location=None):
         return torch.load(f, map_location=map_location)
 
 
-def get_filesystem(path: Union[str, Path]):
+def get_filesystem(path: Union[str, Path]) -> LocalFileSystem:
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/cloud_io.py:39]: This can be something other than `LocalFileSystem`, a better return type would be `AbstractFileSystem`: https://github.com/intake/filesystem_spec/blob/bdcdb304bea87c2d458c6045cfbf28d963fa715b/fsspec/spec.py#L97

**Review by ethanwharris — APPROVED** [review]
Awesome! LGTM :smiley:

**Review by awaelchli — APPROVED** [review]
that was a bit large ^^

File: pytorch_lightning/utilities/apply_func.py [hunk 8]
```diff
@@ -56,11 +55,11 @@ def from_numpy(value, device: torch.device = None):
 
 def apply_to_collection(
     data: Any,
-    dtype: Union[type, tuple],
+    dtype: Union[torch.dtype, Tuple[torch.dtype]],
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/apply_func.py]: Yes, I agree `type` should be used here instead of `torch.dtype`

File: pytorch_lightning/utilities/cli.py [hunk 9]
```diff
@@ -50,9 +51,9 @@ def __init__(self, *args, parse_as_dict: bool = True, **kwargs) -> None:
 
     def add_lightning_class_args(
         self,
-        lightning_class: Union[Type[Trainer], Type[LightningModule], Type[LightningDataModule]],
+        lightning_class: Union['pl.Trainer', 'pl.LightningModule', 'pl.LightningDataModule'],
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/cli.py]: Yes, all the cases should be prepended with `Type`. I forgot to add them back. Thanks

File: pytorch_lightning/utilities/cli.py [hunk 10]
```diff
@@ -92,18 +93,18 @@ class LightningCLI:
...
         self,
-        model_class: Type[LightningModule],
-        datamodule_class: Type[LightningDataModule] = None,
+        model_class: Type['pl.LightningModule'],
+        datamodule_class: Optional[Type['pl.LightningDataModule']] = None,
         save_config_callback: Type[SaveConfigCallback] = SaveConfigCallback,
-        trainer_class: Type[Trainer] = Trainer,
-        trainer_defaults: Dict[str, Any] = None,
-        seed_everything_default: int = None,
+        trainer_class: Type['pl.Trainer'] = Trainer,
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/cli.py:101]: Thank you for catching! :)

File: pytorch_lightning/utilities/memory.py [hunk 11]
```diff
@@ -13,11 +13,15 @@
...
 
 import gc
+from typing import Any, Dict, Union
 
 import torch
 
 
-def recursive_detach(in_dict: dict, to_cpu: bool = False) -> dict:
+def recursive_detach(
+    in_dict: Dict[Any, Union[torch.Tensor, Dict[Any, torch.Tensor]]],
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/memory.py]: I tried to propose a solution to reflect typing for recursive functions.

File: pytorch_lightning/utilities/apply_func.py [hunk 12]
```diff
@@ -161,14 +160,11 @@ def batch_to(data):
     return apply_to_collection(batch, dtype=dtype, function=batch_to)
 
 
-def convert_to_tensors(data, device: torch.device = None):
-    if device is None:
-        raise MisconfigurationException("device (torch.device) should be provided.")
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:166]: why is this removed?

**Comment by Borda** [comment]
there are still some remaining issues:
```
pytorch_lightning/utilities/metrics.py:22: error: Cannot resolve name "RECURSIVE_DICT_WITH_TENSORS" (possible cyclic definition)  [misc]
pytorch_lightning/utilities/metrics.py:44: error: Incompatible return value type (got "Dict[str, Union[Dict[str, float], Any]]", expected "Dict[str, float]")  [return-value]
pytorch_lightning/utilities/memory.py:20: error: Cannot resolve name "RECURSIVE_DICT_WITH_TENSORS" (possible cyclic definition)  [misc]
pytorch_lightning/utilities/enums.py:30: error: Argument 1 of "__eq__" is incompatible with supertype "str"; supertype defines the argument type as "object"  [override]
pytorch_lightning/utilities/enums.py:30: note: This violates the Liskov substitution principle
pytorch_lightning/utilities/enums.py:30: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
pytorch_lightning/utilities/enums.py:30: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
pytorch_lightning/utilities/enums.py:30: note:     def __eq__(self, other: object) -> bool:
pytorch_lightning/utilities/enums.py:30: note:         if not isinstance(other, LightningEnum):
pytorch_lightning/utilities/enums.py:30: note:             return NotImplemented
pytorch_lightning/utilities/enums.py:30: note:         return <logic to compare two LightningEnum instances>
pytorch_lightning/utilities/enums.py:30: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"  [override]
pytorch_lightning/utilities/xla_device.py:23: error: Cannot determine type of '_XLA_AVAILABLE'  [has-type]
pytorch_lightning/utilities/xla_device.py:43: error: Need type annotation for 'queue'  [var-annotated]
pytorch_lightning/utilities/distributed.py:34: error: Name 'ReduceOp' already defined (possibly by an import)  [no-redef]
pytorch_lightning/utilities/distributed.py:37: error: Name 'group' already defined (possibly by an import)  [no-redef]
pytorch_lightning/utilities/distributed.py:47: error: Missing return statement  [return]
pytorch_lightning/utilities/distributed.py:83: error: Argument 1 to "partial" has incompatible type "Optional[Callable[..., Any]]"; expected "Callable[..., Any]"  [arg-type]
pytorch_lightning/utilities/distributed.py:177: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/distributed.py:298: error: "None" not callable  [misc]
pytorch_lightning/utilities/distributed.py:304: error: "None" not callable  [misc]
pytorch_lightning/utilities/distributed.py:306: error: "None" not callable  [misc]
pytorch_lightning/utilities/distributed.py:311: error: "None" not callable  [misc]
pytorch_lightning/utilities/distributed.py:311: error: Item "None" of "Optional[Callable[..., Any]]" has no attribute "__qualname__"  [union-attr]
pytorch_lightning/utilities/apply_func.py:89: error: Argument 2 to "isinstance" has incompatible type "_SpecialForm"; expected "Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]"  [arg-type]
pytorch_lightning/utilities/apply_func.py:118: error: Invalid type comment or annotation  [valid-type]
pytorch_lightning/utilities/apply_func.py:118: note: Suggestion: use type[...] instead of type(...)
pytorch_lightning/utilities/apply_func.py:160: error: Argument "dtype" to "apply_to_collection" has incompatible type "object"; expected "Union[type, Tuple[type]]"  [arg-type]
pytorch_lightning/utilities/apply_func.py:165: error: Argument 1 to "partial" has incompatible type "object"; expected "Callable[..., <nothing>]"  [arg-type]
pytorch_lightning/utilities/warnings.py:22: error: Need type annotation for 'warnings' (hint: "warnings: Set[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/warnings.py:27: error: "None" not callable  [misc]
pytorch_lightning/utilities/debugging.py:35: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/debugging.py:35: error: Missing return statement  [return]
pytorch_lightning/utilities/debugging.py:47: error: Need type annotation for 'logged_metrics' (hint: "logged_metrics: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:48: error: Need type annotation for 'pbar_added_metrics' (hint: "pbar_added_metrics: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:49: error: Need type annotation for 'saved_train_losses' (hint: "saved_train_losses: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:50: error: Need type annotation for 'saved_val_losses' (hint: "saved_val_losses: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:51: error: Need type annotation for 'saved_test_losses' (hint: "saved_test_losses: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:52: error: Need type annotation for 'early_stopping_history' (hint: "early_stopping_history: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:53: error: Need type annotation for 'checkpoint_callback_history' (hint: "checkpoint_callback_history: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:54: error: Need type annotation for 'events' (hint: "events: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:55: error: Need type annotation for 'saved_lr_scheduler_updates' (hint: "saved_lr_scheduler_updates: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:56: error: Need type annotation for 'train_dataloader_calls' (hint: "train_dataloader_calls: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:57: error: Need type annotation for 'val_dataloader_calls' (hint: "val_dataloader_calls: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:58: error: Need type annotation for 'test_dataloader_calls' (hint: "test_dataloader_calls: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:59: error: Need type annotation for 'dataloader_sequence_calls' (hint: "dataloader_sequence_calls: List[<type>] = ...")  [var-annotated]
pytorch_lightning/utilities/debugging.py:189: error: Item "None" of "Optional[ModelCheckpoint]" has no attribute "monitor"  [union-attr]
pytorch_lightning/utilities/debugging.py:202: error: Need type annotation for 'counts'  [var-annotated]
pytorch_lightning/utilities/debugging.py:210: error: Need type annotation for 'counts'  [var-annotated]
pytorch_lightning/utilities/parsing.py:101: error: "None" not callable  [misc]
pytorch_lightning/utilities/parsing.py:105: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/parsing.py:121: error: Missing return statement  [return]
pytorch_lightning/utilities/parsing.py:126: error: Argument 1 to "_get_first_if_any" has incompatible type "List[Parameter]"; expected "Sequence[Namespace]"  [arg-type]
pytorch_lightning/utilities/parsing.py:126: error: Argument 2 to "_get_first_if_any" has incompatible type "Literal[_ParameterKind.VAR_POSITIONAL]"; expected "Type[Any]"  [arg-type]
pytorch_lightning/utilities/parsing.py:127: error: Argument 1 to "_get_first_if_any" has incompatible type "List[Parameter]"; expected "Sequence[Namespace]"  [arg-type]
pytorch_lightning/utilities/parsing.py:127: error: Argument 2 to "_get_first_if_any" has incompatible type "Literal[_ParameterKind.VAR_KEYWORD]"; expected "Type[Any]"  [arg-type]
pytorch_lightning/utilities/parsing.py:133: error: Argument 1 to "getargvalues" has incompatible type "object"; expected "FrameType"  [arg-type]
pytorch_lightning/utilities/parsing.py:162: error: Argument 1 to "getargvalues" has incompatible type "object"; expected "FrameType"  [arg-type]
pytorch_lightning/utilities/parsing.py:200: error: Item "None" of "Optional[FrameType]" has no attribute "f_back"  [union-attr]
pytorch_lightning/utilities/parsing.py:284: error: Argument 1 to "append" of "list" has incompatible type "Union[Dict[Any, Any], Namespace]"; expected "LightningModule"  [arg-type]
pytorch_lightning/utilities/seed.py:53: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "Optional[int]")  [assignment]
pytorch_lightning/utilities/seed.py:54: error: Argument 1 to "int" has incompatible type "Optional[int]"; expected "Union[str, bytes, SupportsInt, _SupportsIndex]"  [arg-type]
pytorch_lightning/utilities/seed.py:57: error: "None" not callable  [misc]
pytorch_lightning/utilities/seed.py:60: error: "None" not callable  [misc]
pytorch_lightning/utilities/device_parser.py:124: error: "None" not callable  [misc]
pytorch_lightning/utilities/device_parser.py:196: error: Argument 2 to "isinstance" has incompatible type "Tuple[Type[List[Any]], _SpecialForm, Type[Set[Any]]]"; expected "Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]"  [arg-type]
pytorch_lightning/utilities/device_parser.py:197: error: Argument 1 to "len" has incompatible type "Union[int, List[int], Tuple[int], Set[int], None]"; expected "Sized"  [arg-type]
pytorch_lightning/utilities/device_parser.py:198: error: Value of type "Union[int, List[int], Tuple[int], Set[int], None]" is not indexable  [index]
pytorch_lightning/utilities/device_parser.py:208: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
pytorch_lightning/utilities/device_parser.py:210: error: Incompatible types in assignment (expression has type "List[int]", variable has type "str")  [assignment]
pytorch_lightning/utilities/device_parser.py:211: error: Incompatible return value type (got "str", expected "Union[int, List[int]]")  [return-value]
pytorch_lightning/utilities/data.py:41: error: "None" not callable  [misc]
pytorch_lightning/utilities/argparse.py:24: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:56: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:81: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:110: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:138: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:148: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
pytorch_lightning/utilities/argparse.py:195: error: Incompatible types in assignment (expression has type "ArgumentParser", variable has type "_ArgumentGroup")  [assignment]
pytorch_lightning/utilities/argparse.py:216: error: Incompatible types in assignment (expression has type "List[type]", variable has type "Tuple[Any, ...]")  [assignment]
pytorch_lightning/utilities/argparse.py:220: error: Need type annotation for 'arg_kwargs' (hint: "arg_kwargs: Dict[<type>, <type>] = ...")  [var-annotated]
pytorch_lightning/utilities/argparse.py:227: error: Incompatible types in assignment (expression has type "Callable[[str], Union[bool, int, str]]", variable has type "Callable[[str], bool]")  [assignment]
pytorch_lightning/utilities/argparse.py:229: error: Incompatible types in assignment (expression has type "Callable[[str], Union[str, bool]]", variable has type "Callable[[str], bool]")  [assignment]
pytorch_lightning/utilities/argparse.py:237: error: Incompatible types in assignment (expression has type "Callable[[str], Union[int, str]]", variable has type "Callable[[str], bool]")  [assignment]
pytorch_lightning/utilities/argparse.py:241: error: Incompatible types in assignment (expression has type "Callable[[Union[int, float, str]], Union[int, float]]", variable has type "Callable[[str], bool]")  [assignment]
pytorch_lightning/utilities/argparse.py:245: error: Incompatible types in assignment (expression has type "Type[float]", variable has type "Callable[[str], bool]")  [assignment]
pytorch_lightning/utilities/argparse.py:259: error: Incompatible return value type (got "_ArgumentGroup", expected "ArgumentParser")  [return-value]
pytorch_lightning/utilities/argparse.py:281: error: Invalid index type "Optional[str]" for "Dict[str, str]"; expected type "str"  [index]
pytorch_lightning/utilities/device_dtype_mixin.py:184: error: Function is missing a type annotation  [no-untyped-def]
pytorch_lightning/trainer/evaluation_loop.py:249: error: "None" not callable  [misc]
pytorch_lightning/accelerators/accelerator.py:490: error: "None" not callable  [misc]
pytorch_lightning/accelerators/accelerator.py:503: error: "None" not callable  [misc]
Found 84 errors in 16 files (checked 408 source files)
```

File: pytorch_lightning/utilities/cli.py [hunk 13]
```diff
@@ -50,9 +51,9 @@ def __init__(self, *args, parse_as_dict: bool = True, **kwargs) -> None:
 
     def add_lightning_class_args(
         self,
-        lightning_class: Union[Type[Trainer], Type[LightningModule], Type[LightningDataModule]],
+        lightning_class: Union[Type['pl.Trainer'], Type['pl.LightningModule'], Type['pl.LightningDataModule']],
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/cli.py]: Thanks for the feedback, I'll take care of it

**Comment by stancld** [comment]
@Borda - Sorry for the late reply. I'm gonna take a look at the remaining issues this weekend.

File: pytorch_lightning/utilities/apply_func.py (hunk 12 above)
Comment (stancld) [reply pytorch_lightning/utilities/apply_func.py:173]: It seemed to me it should work fine with `device=None`. But maybe I was a bit overactive and setting `device` args as non-optional with a default value `None` and raising this message when `device is None` is more appropriate. I can change that back.

File: pytorch_lightning/utilities/parsing.py [hunk 14]
```diff
@@ -185,11 +195,11 @@ def flatten_dict(source, result=None):
...
 def save_hyperparameters(
     obj: Any,
-    *args,
+    *args: Any,
     ignore: Optional[Union[Sequence[str], str]] = None,
-    frame: Optional[types.FrameType] = None
+    frame: Optional[types.FrameType] = None,
 ) -> None:
-    """See :meth:`~pytorch_lightning.LightningModule.save_hyperparameters`"""
+    """See :meth:`~pytorch_lightning.'pl.core.lightning.LightningModule'.save_hyperparameters`"""
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:202]:
```suggestion
    """See :meth:`~pytorch_lightning.core.lightning.LightningModule'.save_hyperparameters`"""
```

File: pytorch_lightning/utilities/parsing.py [hunk 15]
```diff
@@ -264,14 +274,14 @@ def __repr__(self):
         return out
 
 
-def _lightning_get_all_attr_holders(model, attribute):
+def _lightning_get_all_attr_holders(model: 'pl.core.lightning.LightningModule', attribute: str) -> Any:
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:277]: Same for the other occurrences in this file

```suggestion
def _lightning_get_all_attr_holders(model: 'pl.LightningModule', attribute: str) -> Any:
```

File: pytorch_lightning/utilities/parsing.py [hunk 16]
```diff
@@ -159,18 +166,21 @@ def collect_init_args(frame, path_args: list, inside: bool = False) -> list:
...
     """
     _, _, _, local_vars = inspect.getargvalues(frame)
-    if '__class__' in local_vars:
-        local_args = get_init_args(frame)
-        # recursive update
-        path_args.append(local_args)
-        return collect_init_args(frame.f_back, path_args, inside=True)
-    elif not inside:
-        return collect_init_args(frame.f_back, path_args, inside)
+    if isinstance(frame.f_back, FrameType):
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:169]: mind comment on why this?

File: pytorch_lightning/utilities/device_parser.py [hunk 17]
```diff
@@ -187,25 +187,30 @@ def _check_data_type(device_ids: Any) -> None:
...
 
-def _tpu_cores_valid(tpu_cores):
+def _tpu_cores_valid(tpu_cores: Optional[Union[int, Iterable[int], List[int], Tuple[int], Set[int]]] = None) -> bool:
     # allow 1 or 8 cores
     if tpu_cores in (1, 8, None):
         return True
 
+    # First condition is necessary for mypy compatibility;
+    # list_tpu_cores is required to declare for mypy compatiblity too
+    if (tpu_cores is not None) and (not isinstance(tpu_cores, int)):
```
Comment (Borda) [inline pytorch_lightning/utilities/device_parser.py:197]:
```suggestion
    if tpu_cores is not None and not isinstance(tpu_cores, int):
```

File: pytorch_lightning/utilities/distributed.py [hunk 18]
```diff
@@ -308,11 +317,13 @@ def register_ddp_comm_hook(
             )
             ddp_comm_hook = ddp_comm_wrapper(ddp_comm_hook)
 
-    rank_zero_debug(f"Registering DDP comm hook: {ddp_comm_hook.__qualname__}.")
-    model.register_comm_hook(
-        state=ddp_comm_state,
-        hook=ddp_comm_hook,
-    )
+    # If condition required for mypy compatibility
+    if ddp_comm_hook is not None:
```
[the remark this answers is not in the document]
Comment (stancld) [reply pytorch_lightning/utilities/distributed.py:321]: If this condition is dropped, the following mypy issue is raised:
```
utilities/distributed.py:320: error: Item "None" of "Optional[Callable[..., Any]]" has no attribute "__qualname__"
```

File: pytorch_lightning/utilities/parsing.py (hunk 16 above)
Comment (stancld) [reply pytorch_lightning/utilities/parsing.py:170]: Dropping this condition leads to the following two issues:
```
utilities/parsing.py:174: error: Argument 1 to "collect_init_args" has incompatible type "Optional[FrameType]"; expected "FrameType"
utilities/parsing.py:176: error: Argument 1 to "collect_init_args" has incompatible type "Optional[FrameType]"; expected "FrameType"
```

Maybe, there's another nice workaround.

**Comment by stancld** [comment]
@Borda Hi, sorry for not responding. This PR looks now obsolete and pretty complicated to merge due to its large extent. Would you mind if I close this PR and maybe open some little PRs for individual files to finish this mypy task?

**Comment by awaelchli** [comment]
If it's possible to break down in smaller pieces that would be awesome! very welcomed!
a smaller, incremental PRs would make it easier to review, lower the changes of conflicts and increase the changes to get merged <3
